### ISSUE TITLE
Add disk/storage from datastore to a VM

### ISSUE THREAD

**Issue by arvindayyangar** [vmware/pyvmomi-community-samples#292]
I want to add storage from a datastore I have created to a VM I have deployed. I was looking at the 'add_disk_to_vm.py' example and it seems to add the disk storage from the datastore the VM is deployed on, is that correct ?

If I understand the code right, I need to get the device information for the datastore I want to add and use that to reconfigure the VM. I looked at the API documentation and could not find a way to get that information for the datastore. 

How can I achieve this ?

**Comment by asonar** [comment]
Below code should help you ....

=============
  def add(vmmor,kwargs={}):
    disknum = 5
    controllerkey = 1000
    for disk in range(disknum):
      dev_changes=[]
      unit_number = 0
      spec = vim.vm.ConfigSpec()
      for dev in vmmor.config.hardware.device:
        if hasattr(dev.backing, 'fileName') and dev.controllerKey == controllerkey:
          unit_number += 1
          if unit_number == 7:
            unit_number += 1
          if unit_number == 16:
            cntr+=1
            if cntr > 3:
              print ("Can not add any more disk to vm : {0}".format(vmmor.name))
              return
            unit_number = 1
            controllerkey +=1
            continue
      disk_spec = vim.vm.device.VirtualDeviceSpec()
      disk_spec.fileOperation = "create"
      disk_spec.operation = vim.vm.device.VirtualDeviceSpec.Operation.add
      disk_spec.device = vim.vm.device.VirtualDisk()
      disk_spec.device.backing = vim.vm.device.VirtualDisk.FlatVer2BackingInfo()
      disk_spec.device.backing.thinProvisioned = True
      disk_spec.device.backing.diskMode = "persistent"
      disk_spec.device.unitNumber = unit_number
      disk_spec.device.capacityInKB = 100 * 1024
      disk_spec.device.controllerKey = controllerkey
      dev_changes.append(disk_spec)
      spec.deviceChange = dev_changes
      vmmor.ReconfigVM_Task(spec=spec)

=============

Is this what you are looking for ?

**Comment by arvindayyangar** [comment]
Thanks asonar.
From the code
> disk_spec.device = vim.vm.device.VirtualDisk()
> disk_spec.device.backing = vim.vm.device.VirtualDisk.FlatVer2BackingInfo()
Isn't this adding the disk form the same datastore where the VM is deployed ?

I might not have been clear in my initial post because I am not very familiar ESX/VMware terminology so let me give an example.
Lets say I have two datastores on my host, ds1 and ds2. I deployed a vm, VM1 on ds1.
When I try to add disk to VM1 using the above API, it is going to carve out space from ds1, is that correct ? How do I add disk space from ds2 ?

**Comment by asonar** [comment]
Yes. this will add the disk from current datastore where VM is placed. I dont have code to select different datastore while adding disk, lemme see if I can get that code for you.

**Comment by arvindayyangar** [comment]
path_on_ds = '[' + datastore + ']' + vm.name
content.fileManager.MakeDirectory(path_on_ds)
ds_obj = get_obj(content, [vim.Datastore], datastore)

**Comment by arvindayyangar** [comment]
I could add disk from a different datastore with the below code changes, hope it helps someone trying to do something similar.      
        path_on_ds = '[' + datastore + ']' + vm.name
        content = si.RetrieveContent()

        content.fileManager.MakeDirectory(path_on_ds)
        ds_obj = get_obj(content, [vim.Datastore], datastore)
        dev_changes = []
        new_disk_kb = int(disk_size) * 1024 * 1024
        disk_spec = vim.vm.device.VirtualDeviceSpec()
        disk_spec.fileOperation = "create"
        disk_spec.operation = vim.vm.device.VirtualDeviceSpec.Operation.add

        disk_spec.device = vim.vm.device.VirtualDisk()
        disk_spec.device.backing = \
            vim.vm.device.VirtualDisk.FlatVer2BackingInfo()
        if disk_type == 'thin':
            disk_spec.device.backing.thinProvisioned = True
        disk_spec.device.backing.diskMode = 'persistent'
        disk_spec.device.backing.fileName = path_on_ds + '/' + vm.name + '_' + str(unit_number) + '.vmdk'
        disk_spec.device.backing.datastore= ds_obj
        disk_spec.device.unitNumber = unit_number
        disk_spec.device.capacityInKB = new_disk_kb
        disk_spec.device.controllerKey = controller.key

**Comment by subhash12** [comment]
I was able to add disk to different datastore.
I had to except from content.fileManager.MakeDirectory, if another disk has to be in same datastore.

Thanks

**Comment by localhostage** [comment]
As of today's latest pull of the SDK @arvindayyangar's post didn't work for me until I added ".vmdk" to the end of `fileName` variable.

Ultimately I ended up with this helper function to add volumes to a VM reference:
```
def add_virtual_disk(vm_ref, size_in_gb, location_datastore):
    unit_number = 0
    for dev in vm_ref.config.hardware.device:
        if hasattr(dev.backing, 'fileName'):
            unit_number = int(dev.unitNumber) + 1
            # unit_number 7 reserved for scsi controller
            if unit_number == 7:
                unit_number += 1
            if unit_number >= 16:
                logging.error("we don't support this many disks")
                return
        if isinstance(dev, vim.vm.device.VirtualSCSIController):
            controller = dev

    path_on_ds = '[' + location_datastore.name + ']' + vm_ref.name
    fileName = path_on_ds + '/' + vm_ref.name + '_' + str(unit_number) + '.vmdk'

    new_disk_kb = int(size_in_gb) * 1024 * 1024
    virtual_disk_backing = vim.vm.device.VirtualDisk.FlatVer2BackingInfo(thinProvisioned=True, diskMode='persistent', fileName=fileName, datastore=location_datastore)
    virtual_disk = vim.vm.device.VirtualDisk(backing=virtual_disk_backing, capacityInKB=new_disk_kb, unitNumber=unit_number, controllerKey=controller.key)
    disk_spec = vim.vm.device.VirtualDeviceSpec(operation='add', fileOperation='create', device=virtual_disk)

    config_spec = vim.vm.ConfigSpec(deviceChange=[disk_spec])
    vm_ref.Reconfigure(spec=config_spec)```

**Comment by bhavsi** [comment]
Is there a REST API way to do this? When adding a disk, it would be added as a separate partition right? Is there a way to add more disks and expand the primary partition?
